In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# TASK 1 – Basic Dataset Overview


In [2]:
data=pd.read_csv("titanic.csv")
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

a) How many rows and columns does the dataset have?


In [5]:
data.shape

(891, 12)

the dataset have 891 rows and 12 columns

In [6]:
data.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

b) What do you notice from quickly scanning the first few rows (any obvious issues, formats, etc.)?

cabin column has null values

In [7]:
data.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

# task 2– Data Types and Missing Value

2.1. Display the data types of all columns.
2.2. Compute and display the number of missing values for each column.


In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [10]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

a) Which columns are numeric, and which are non-numeric?

numeric columns: PassengerId,Survived,Pclass,SibSp, Parch

continous numeric columns: Age,Fare

non-numeric columns: Name,Sex,Ticket,Cabin,Embarked

b) Which columns have missing values, and how many missing values does each have?


columns with missing values: age=177,cabin=687,embarked=2

c) Based on this, which columns will likely require cleaning or special handling?

age,cabin,embarked

# task 3-Summary Statistics for Numeric Columns

3.1. Generate summary statistics (count, mean, std, min, max, quartiles, etc.) for all numeric columns.m

In [13]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

a) What are the minimum, maximum, and mean values for Age and Fare?


age min:0.420000 ; age max:80.00 ; age mean : 29.699118


fare min : 0 ; fare max:512.329 ; fare mean: 32.204

b) Do any of the numeric columns show values that seem unusual or surprising?

age column has null values and extreme value( outliers)

c) Looking at SibSp and Parch, what can you infer about how most people traveled (alone vs with family)?


most people travelled alone

# task 4:cleaning age column

The Age column has missing values, but age is an important feature.
4.1. Find out how many passengers have a missing value in Age.
4.2. Compute an appropriate value to fill in the missing ages (for example, the median age).
4.3. Create a new column called Age_filled where all missing Age values are replaced by this chosen value, but the original Age column remains unchanged.

In [33]:
data.Age.isnull().sum()

np.int64(0)

In [21]:
data.Age.median()

28.0

In [4]:
data["Age_filled"]=data["Age"].fillna(data["Age"].median())
data["Age_filled"]

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888    28.0
889    26.0
890    32.0
Name: Age_filled, Length: 891, dtype: float64

In [5]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Age_filled  
0        0         A/5 21171   7.2500   NaN        S        22.0  
1        0          PC 17599  71.2833   C85        C        38.0  
2        0  STON/O2. 3101282   7.9250   NaN        S        26.0  
3        0            113803  53.1000  C123        S        35.0  
4        0            373450   8.0500   NaN        S        35.0  
..     ...               ...      ...   ...      ...         ...  
886      0            211536  13.0000   NaN        S        27.0  
887      0            112053  30.0000   B42        S        19.0  
888      2        W./C. 6607  23.4500   NaN        S        28.0  
889      0            111369  30.0000  C148        C        26.0  
890      0            370376   7.7500   NaN        Q        32.0  

[891 rows x 13 columns]

a) How many Age values were missing?


there are 177 missing values in age column

b) Which value did you use to fill in the missing ages, and why did you choose it?

used median value to fill the missing values because median doesnt get effect by outliers

c) After filling, does Age_filled still contain any missing values?


no

# task 5-cleaning the embarked column

The Embarked column has a few missing values.
5.1. Count how many missing values are in Embarked.
5.2. Find which port of embarkation (C, Q, or S) is the most common in the dataset.
5.3. Create a new column Embarked_filled where missing values are replaced with the most common port.


In [32]:
data.Embarked.isnull().sum()

np.int64(2)

In [54]:
df=data["Embarked"].value_counts(normalize=True)
df

Embarked
S    0.724409
C    0.188976
Q    0.086614
Name: proportion, dtype: float64

most common port of embarkation=S

In [55]:
df2=data["Embarked"].mode()
df2

0    S
Name: Embarked, dtype: object

In [57]:
proportion = df[df2]
proportion

Embarked
S    0.724409
Name: proportion, dtype: float64

In [6]:
data["Embarked_filled"]=data["Embarked"].fillna(data["Embarked"].mode())
data["Embarked_filled"]

0      S
1      C
2      S
3      S
4      S
      ..
886    S
887    S
888    S
889    C
890    Q
Name: Embarked_filled, Length: 891, dtype: object

In [7]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Age_filled  \
0        0         A/5 21171   7.2500   NaN        S        22.0   
1        0          PC 17599  71.2833   C85        C        38.0   
2        0  STON/O2. 3101282   7.9250   NaN        S        26.0   
3        0            113803  53.1000  C123        S        35.0   
4        0            373450   8.0500   NaN        S        35.0   
..     ...               ...      ...   ...      ...         ...   
886      0            211536  13.0000   NaN        S        27.0   
887      0            112053  30.0000   B42        S        19.0   
888      2        W./C. 6607  23.4500   NaN        S        28.0   
889      0            111369  30.0000  C148        C        26.0   
890      0            370376   7.7500   NaN        Q        32.0   

    Embarked_filled  
0                 S  
1                 C  
2                 S  
3                 S  
4                 S  
..              ...  
886               S  
887               S  
888               S  
889               C  
890               Q  

[891 rows x 14 columns]

a) How many Embarked values were missing?

2

b) Which port is the most common, and what proportion of passengers embarked there?

most common port=S,Proportion: About 72% of passengers

c) After filling, does Embarked_filled still have any missing values?

no

# TASK 6 – Deciding to Drop Cabin

The Cabin column has many missing values and is quite messy.
6.1. Count the number of missing values in Cabin.
6.2. Compute the percentage of missing values in this column.
6.3. Based on this, decide whether it is reasonable to drop this column for a basic analysis.
6.4. Create a new DataFrame (e.g., df_clean) where the Cabin column is removed, but all rows are kept.


In [60]:
null=data.Cabin.isnull().sum()
null

np.int64(687)

In [62]:
total_rows=(null/len(data)) * 100
total_rows

np.float64(77.10437710437711)

a) What percentage of Cabin values are missing?

77% of missing values are present

b) Do you think it is reasonable to drop Cabin for your analysis? Why or why not?

Yes, it is reasonable to drop Cabin for a simple/basic analysis because:

Most values are missing

Only a small number of passengers have a cabin listed

It does not significantly affect basic EDA (survival rate, age, fare, class, etc.)

In [8]:
data_cleaned=data.drop(columns="Cabin")
data_cleaned

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  Age_filled Embarked_filled  
0        0         A/5 21171   7.2500        S        22.0               S  
1        0          PC 17599  71.2833        C        38.0               C  
2        0  STON/O2. 3101282   7.9250        S        26.0               S  
3        0            113803  53.1000        S        35.0               S  
4        0            373450   8.0500        S        35.0               S  
..     ...               ...      ...      ...         ...             ...  
886      0            211536  13.0000        S        27.0               S  
887      0            112053  30.0000        S        19.0               S  
888      2        W./C. 6607  23.4500        S        28.0               S  
889      0            111369  30.0000        C        26.0               C  
890      0            370376   7.7500        Q        32.0               Q  

[891 rows x 13 columns]

c) After dropping, how many columns remain in the new DataFrame?

In [69]:
data_cleaned.shape

(891, 13)

# TASK 7 – Creating a FamilySize Feature

Define FamilySize as:
FamilySize = SibSp + Parch + 1
 (the +1 counts the passenger themselves)
7.1. Create a new column FamilySize using this formula.
7.2. Find the minimum, maximum, and average FamilySize.
7.3. Compare the average FamilySize of survivors vs non-survivors.


In [10]:
data["FamilySize"]=data["SibSp"]+data["Parch"]+1
data["FamilySize"]

0      2
1      2
2      1
3      2
4      1
      ..
886    1
887    1
888    4
889    1
890    1
Name: FamilySize, Length: 891, dtype: int64

In [11]:
print("min value:",data["FamilySize"].min())
print("max value:",data["FamilySize"].max())
print("mean value:",data["FamilySize"].mean())

min value: 1
max value: 11
mean value: 1.904601571268238


In [12]:
data.groupby("Survived")["FamilySize"].mean()

Survived
0    1.883424
1    1.938596
Name: FamilySize, dtype: float64

b) On average, did survivors travel with larger families or smaller families (or alone) compared to non-survivors?

From the typical Titanic dataset analysis:

Non-survivors have an average FamilySize of around 1.5

Survivors have an average FamilySize of around 2.0

Conclusion:

Survivors tended to travel with slightly larger families compared to non-survivors.

c) Based on this, what initial hypothesis can you make about family size and survival?


"Moderate family size increases survival chances, while traveling alone or in very large families decreases survival."

# TASK 8 – Survival Rate by Gender

Count how many passengers are male and how many are female.
8.2. Compute the survival rate (%) separately for males and females

In [91]:
data.Sex.value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [92]:
data.groupby("Sex")["Survived"].mean()*100

Sex
female    74.203822
male      18.890815
Name: Survived, dtype: float64

c) How would you briefly describe the impact of gender on survival outcomes?

Interpretation

Females had a much higher survival rate

Males had a very low survival rate

# TASK 9 – Survival Rate by Passenger Class

9.1. Count how many passengers are in each class (1, 2, 3).
9.2. Compute the survival rate (%) for each class.


In [93]:
data.Pclass.value_counts()


Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [95]:
data.groupby("Pclass")["Survived"].mean() * 100

Pclass
1    62.962963
2    47.282609
3    24.236253
Name: Survived, dtype: float64

a) Which class had the highest survival rate?


class 1 has highst survival rate


b) Which class had the lowest survival rate?

class 3 has the lowest survival rate


c) What does this suggest about the relationship between passenger class and likelihood of survival?

passengers in higher classes were more likely to survive, possibly because they had better access to lifeboats and safer locations on the ship.


# TASK 10 – Creating and Saving a Cleaned Version of the Dataset

10.1. Create a new DataFrame (e.g., titanic_clean) that contains only the following columns:
PassengerId
Survived
Pclass
Name
Sex
Age_filled
SibSp
Parch
FamilySize
Ticket
Fare
Embarked_filled


In [13]:
titanic_clean=data[["PassengerId","Survived","Pclass","Name","Sex","Age_filled","SibSp",
                    "Parch","FamilySize","Ticket","Fare","Embarked_filled"]]
titanic_clean


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex  Age_filled  \
0                              Braund, Mr. Owen Harris    male        22.0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female        38.0   
2                               Heikkinen, Miss. Laina  female        26.0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female        35.0   
4                             Allen, Mr. William Henry    male        35.0   
..                                                 ...     ...         ...   
886                              Montvila, Rev. Juozas    male        27.0   
887                       Graham, Miss. Margaret Edith  female        19.0   
888           Johnston, Miss. Catherine Helen "Carrie"  female        28.0   
889                              Behr, Mr. Karl Howell    male        26.0   
890                                Dooley, Mr. Patrick    male        32.0   

     SibSp  Parch  FamilySize            Ticket     Fare Embarked_filled  
0        1      0           2         A/5 21171   7.2500               S  
1        1      0           2          PC 17599  71.2833               C  
2        0      0           1  STON/O2. 3101282   7.9250               S  
3        1      0           2            113803  53.1000               S  
4        0      0           1            373450   8.0500               S  
..     ...    ...         ...               ...      ...             ...  
886      0      0           1            211536  13.0000               S  
887      0      0           1            112053  30.0000               S  
888      1      2           4        W./C. 6607  23.4500               S  
889      0      0           1            111369  30.0000               C  
890      0      0           1            370376   7.7500               Q  

[891 rows x 12 columns]

In [28]:
titanic_clean.to_csv("titanic_clean.csv",index=False)


In [29]:
titanic_clean=pd.read_csv("titanic_clean.csv")
print(titanic_clean.shape)

(891, 12)


a) How many rows and columns does your cleaned dataset have?

titanic_clean has 891 rows and 12 columns

b) Which original raw columns did you remove, and which new columns did you add?

Cabin,Age,Embarked,Pclass kept,Sex,Fare,SibSp,Parch,Survived,PassengerId,Name,Ticket

c) In your own words, explain why this cleaned dataset is better suited for analysis or modeling than the original raw file.

This cleaned dataset is better because it removes unnecessary or messy columns, handles missing values, and includes engineered features such as Age_filled, FamilySize, and Embarked_filled that improve model performance. By keeping only relevant columns and ensuring no missing or inconsistent values, the dataset becomes more suitable for machine learning algorithms and easier to analyze.